File: Erlang/hello_world/src/extra/road.erl
-module(road).
-compile(export_all).

-spec main() -> [{atom(), integer()}].
main() ->
    File = "road.txt",
    {ok, Binary} = file:read_file(File),
    optimal_path(parse_map(Binary)).

%% Transform a string into a readable map of triples
-spec parse_map(binary() | string()) -> [{integer(), integer(), integer()}].
parse_map(Binary) when is_binary(Binary) ->
    parse_map(binary_to_list(Binary));
parse_map(String) when is_list(String) ->
    Values = [list_to_integer(X) || X <- string:tokens(String, "\r\n\t ")],
    group_vals(Values, []).

-spec group_vals([integer()], [{integer(), integer(), integer()}]) -> [{integer(), integer(), integer()}].
group_vals([], Acc) ->
    lists:reverse(Acc);
group_vals([A, B, X | Rest], Acc) ->
    group_vals(Rest, [{A, B, X} | Acc]).

%% actual problem solving
%% change triples of the form {A,B,X}
%% where A,B,X are distances and a,b,x are possible paths
%% to the form {DistanceSum, PathList}.
-spec shortest_steps({integer(), integer(), integer()}, {{integer(), [{atom(), integer()}]}, {integer(), [{atom(), integer()}]}}) -> {{integer(), [{atom(), integer()}]}, {integer(), [{atom(), integer()}]}}.
shortest_steps({A, B, X}, {{DistA, PathA}, {DistB, PathB}}) ->
    OptA1 = {DistA + A, [{a, A} | PathA]},
    OptA2 = {DistB + B + X, [{x, X}, {b, B} | PathB]},
    OptB1 = {DistB + B, [{b, B} | PathB]},
    OptB2 = {DistA + A + X, [{x, X}, {a, A} | PathA]},
    {erlang:min(OptA1, OptA2), erlang:min(OptB1, OptB2)}.

%% Picks the best of all paths, woo!
-spec optimal_path([{integer(), integer(), integer()}]) -> [{atom(), integer()}].
optimal_path(Map) ->
    {A, B} = lists:foldl(fun shortest_steps/2, {{0, []}, {0, []}}, Map),
    {_Dist, Path} = if hd(element(2, A)) =/= {x, 0} -> A;
                       hd(element(2, B)) =/= {x, 0} -> B
    end,
    lists:reverse(Path).

File: Elixir/hello_world_again/lib/extra/road.ex
defmodule Road do
  def run() do
    file = "road.txt"
    {:ok, binary} = File.read(file)
    parse_map(binary)
    |> group_values([])
    |> optimal_path()
  end

  @spec parse_map(binary) :: [integer()]
  def parse_map(string) do
    String.split(string, ["\r", "\n", "\t", " "], trim: true)
    |> Enum.map(&String.to_integer/1)
  end

  def group_values([], acc), do: Enum.reverse(acc)
  def group_values([a, b, x | xs], acc) do
    group_values(xs, [{a, b, x} | acc])
  end

  def shortest_steps(elem, acc) do
    {a, b, x} = elem
    { {dist_a, path_a}, {dist_b, path_b} } = acc
    opt_a1 = { dist_a + a, [{:a, a} | path_a] }
    opt_a2 = { dist_b + b + x, [{:x, x}, {:b, b} | path_b] }
    opt_b1 = { dist_b + b, [{:b, b} | path_b] }
    opt_b2 = { dist_a + a + x, [{:x, x}, {:a, a} | path_a] }
    { min(opt_a1, opt_a2), min(opt_b1, opt_b2) }
  end

  def optimal_path(values) do
    { a, b } = Enum.reduce(values, { {0, []}, {0, []} }, &shortest_steps/2)
    { _, path } = if hd(elem(a, 1)) != {:x, 0} do
      a
    else
      if hd(elem(b, 1)) != {:x, 0} do
        b
      else
        { 0, [] }
      end
    end
    Enum.reverse(path)
  end
end
